docs/src/main/asciidoc/tests-with-coverage-guide.adoc

```diff
@@ -0,0 +1,701 @@
+
+During the execution of the generation of the report, you may have noticed a warning:
+
+[source,shell,subs=attributes+]
+----
+[WARNING] Classes in bundle '***' do no match with execution data. For report generation the same class files must be used as at runtime.
+[WARNING] Execution data for class org/acme/quickstart/GreetingResource does not match.
+----
+
+It seems that Quarkus and JaCoCo step on each other's toes. What happens is that Quarkus transforms the JAX-RS resources (and also the Panache files).
+You may have noticed that `GreetingResource` was not written in the simplest way like:
+
+[source,java]
+----
+...
+@Path("/hello")
+public class GreetingResource {
+
+    @Inject
+    GreetingService service;
+
+    @GET
+    @Produces(MediaType.TEXT_PLAIN)
+    @Path("/greeting/{name}")
+    public String greeting(@PathParam("name") String name) {
+        return service.greeting(name);
+    }
+
+    @GET
+    @Produces(MediaType.TEXT_PLAIN)
+    public String hello() {
+        return "hello";
+    }
+}
+----
+
+Above, the constructor is implicit and we use injection to have an instance of `GreetingService`. Note that, with this code relying on an implicit constructor, the coverage would have been reported properly by JaCoCo.
+Instead, we introduced a constructor based injection:
+
+[source,java]
+----
+...
+@Path("/hello")
+public class GreetingResource {
+
+    private final GreetingService service;
+
+    @Inject
+    public GreetingResource(GreetingService service) {
+        this.service = service;
+    }
+...
+}
+----
+
+Some might say that this approach is preferable since the field can be *final* like this. Anyway, in some cases you might need an explicit constructor. And, in that case, the coverage is not reported properly by JaCoCo.
+This is because Quarkus generates a constructor without any parameter and does some bycode manipulations in order to add it to the class. It is what happened here, just before the execution of the integration tests:
+
+[source,shell,subs=attributes+]
+----
+[INFO] --- quarkus-maven-plugin:0.16.0:build (default) @ getting-started-testing ---
+[INFO] [io.quarkus.deployment.QuarkusAugmentor] Beginning quarkus augmentation
+...
+----
+
+As a consequence, JaCoCo does not recognize the classes when it wants to create its report. But wait... there is a solution.
+
+== Instrumenting the classes instead
+
+JaCoCo has two modes. The first one is based on an agent and instruments classes on-the-fly. Unfortunately, this is incompatible with the dynamic classfile transformations that Quarkus does. But there is a second mode called link:https://www.eclemma.org/jacoco/trunk/doc/offline.html[offline instrumentation]. Classes are pre-instrumented in advance via the *jacoco:instrument* Maven goal and during their usage (when the tests are ran), *jacocoagent.jar* must be added in the classpath.
```

Comment:
"But there is a second mode" -> "The second mode is"

Reply:
"must be added in the classpath." - > "must be added to the classpath."

Reply:
Changed